modules/custom/activity_send/modules/activity_send_email/src/Plugin/AdvancedQueue/JobType/ActivitySendEmailJobType.php

```diff
@@ -0,0 +1,344 @@
+          $this->getLogger('activity_send_email_worker')->notice('The activity was already deleted. We marked it as successful.');
+          return JobResult::success('The activity was already deleted. We marked it as successful.');
+        }
+
+        // Is the website multilingual.
+        $is_multilingual = $this->languageManager->isMultilingual();
+
+        if (empty($data['recipients'])) {
+          $recipients = array_column($activity->field_activity_recipient_user->getValue(), 'target_id');
+
+          if (count($recipients) > 50) {
+            if ($is_multilingual) {
+              // We also want to send emails to users per language in a given
+              // frequency.
+              foreach ($languages = $this->languageManager->getLanguages() as $language) {
+                $langcode = $language->getId();
+                // Load all user by given language.
+                $user_ids_per_language = $this->database->select('users_field_data', 'ufd')
+                  ->fields('ufd', ['uid'])
+                  ->condition('uid', $recipients, 'IN')
+                  ->condition('preferred_langcode', $langcode)
+                  ->execute()->fetchAllKeyed(0, 0);
+
+                // Prepare the batch per language.
+                $this->prepareBatch($data, $user_ids_per_language, $langcode);
+              }
+            }
+            else {
+              $this->prepareBatch($data, $recipients);
+            }
+            // We split up in batches. We can stop processing this specific queue
+            // item.
+            $this->getLogger('activity_send_email_worker')->notice('The Job was not finished correctly, no error thrown.');
+            return JobResult::success('The Job was has been split up in batches.');
+          }
+        }
+        else {
+          $recipients = $data['recipients'];
+        }
+
+        if (!empty($recipients)) {
+          // Get Message Template id.
+          $message_storage = $this->entityTypeManager->getStorage('message');
+          /** @var \Drupal\message\Entity\Message $message */
+          $message = $message_storage->load($activity->getFieldValue('field_activity_message', 'target_id'));
+          $message_template_id = $message->getTemplate()->id();
+
+          // Prepare an array of all details required to process the item.
+          $parameters = [
+            'activity' => $activity,
+            'message' => $message,
+            'message_template_id' => $message_template_id,
+          ];
+
+          // We want to give preference to users who have set notification
+          // settings as 'immediately'.
+          $email_frequencies = [
+            'immediately',
+            'daily',
+            'weekly',
+            'none',
+          ];
+
+          // Let's store the users IDs which will be processed by the loop.
+          $processed_users = [];
+          foreach ($email_frequencies as $email_frequency) {
```

Comment:
This section of code is also being used in `ActivitySendEmailWorker` class. Shall we move this code to a common service?

Reply:
Or, as we will be removing that class anyway in 11.x, we can also leave it like this.